index.bs

```diff
@@ -309,7 +309,9 @@ Every {{Document}} has a <dfn for="Document">post-prerendering activation steps
 
       1. Set |successorBC|'s [=browsing context/loading mode=] to "`default`".
 
-      1. [=Fire an event=] named {{Document/prerenderingchange}} at |successorBC|'s [=active document=].
+      1. Let |d| be |successorBC|'s [=active document=].
+
+      1. [=Fire an event=] named {{Document/prerenderingchange}} at |d| and at each [=active document=] of each of |d|'s [=list of the descendant browsing contexts=].
```

Comment:
It's not quite obvious from this phrasing that it's fired in order. I'd prefer "For each descendantBC of d's list of descendant browsing contexts, fire an event named prerenderingchange at descendantBC's active document."

Also, should we queue a separate task for each of these? That seems potentially necessary since the other BCs could be cross-process?

Reply:
Done. And yes, I think separate tasks makes sense, though I only have a cursory knowledge about the details there so I just cargo cult-ed from above. Does this look ok?

Reply:
Hmm, will any descendant browsing contexts be cross-process? My understanding is that we'll only load same-origin iframes whilst prerendering, in which case I think they are all in the same agent cluster and BCG? Does that impact any of this?

Reply:
I don't see why we couldn't load cross-origin subframes if they were themselves storageless or opted in. I don't think we should bake in an assumption that all nested BCs share an event loop, so queuing tasks makes sense to me.

Reply:
Maybe not (I'm not caught up on that discussion), but it seems like something that could change? This still works in the case where we don't load cross-process iframes so it seems like a better approach to me. The only drawback I can see is that subframe's may still show prerendering state while we flush the `post-prerendering activation steps list` of the current document. That seems probably ok to me since this list is associated with specific document.

Reply:
That said this current description doesn't make it clear to me if these other contexts see `document.prerendering` change at the same time as they receive the event, which they should. This may require us to explicitly plumb a flag through to them. We probably should do this anyway, as synchronously asking for a property of a browsing context with another event loop (as we do here) is kinda a dodge.

Reply:
That's a good point. However, I'm generally a bit scared about posting tasks into other globals from one global. So I slightly prefer individual tasks.

@bokand, how do you plan to implement this? Maybe that can be the tiebreaker.

Reply:
Oh, I see people were talking while I was reviewing. Yeah, I agree with @jeremyroman's concern that the event and `document.prerendering` updating will not necessarily happen at the same time with this spec text.

Reply:
Yeah so just so that I'm clear, I think the concern here is that documents in the same agent that indeed share the same event loop and thus can synchronously access each other, have their `document.prerendering` updated at different times. Am I right in thinking that is the concern?

Of course documents across agent clusters will necessarily have their `document.prerendering` state updated at different times.

Reply:
Hmm, that might be another concern. But Jeremy's concern is that as defined, the `prerendering` getter updates instantly (since the TLBC switches from prerendering to not-prerendering at some point in time). Whereas the event fires from a queued task.

I guess that's not so bad, compared to the reverse...

Reply:
Ah, I see. One consequence of that is since we're planning on using the state of the TLBC when considering fetching restrictions for a given request, then it is possible that a web page could observe that a request was sent with no fetching restrictions _just shortly_ before the `prerendering` getting changes its value, but I agree that doesn't seem like a big deal.

My concern is that multiple documents sharing the same agent might be able to observe an inconsistency in their `prerendering` getter values, which really isn't a big deal, but I just want to make sure that it is known.

Reply:
Thanks all.

Agree `document.prerendering` shouldn't change until the event is dispatched so I'll go ahead and fix that.

Yeah, this does mean that the `prerendering` value may be inconsistent between documents but I don't think that should be an issue. Especially since, within an agent, the ordering will be consistent (parent frames will always make and see the change before children). The expected use cases should really only be checking their own `document`'s value anyway.

Just to clarify:

> That's a good point. However, I'm generally a bit scared about posting tasks into other globals from one global. So I slightly prefer individual tasks.

I wasn't sure who this was in reply to, what do you mean by "posting tasks into other globals from one global" vs "individual tasks"? Is the current PR "posting into other globals" (since each posted task uses a different `window`?)

> @bokand, how do you plan to implement this? Maybe that can be the tiebreaker.

I hadn't thought that far ahead :). The prerequisite CL plumbing this state into the renderer just landed though so I'll go ahead and try this out before making further progress here.

Reply:
> I wasn't sure who this was in reply to, what do you mean by "posting tasks into other globals from one global" vs "individual tasks"? Is the current PR "posting into other globals" (since each posted task uses a different window?)

I phrased this poorly. The current PR is "individual tasks".

The alternative I'm contrasting it with is posting a single task, from the parent global, which updates the prerendering state of all the same-agent child globals. There's nothing really wrong with that, but it's a bit scary to me.